tests/integration/api/mutations/taxRates/taxRates.test.js

```diff
@@ -0,0 +1,213 @@
+import encodeOpaqueId from "@reactioncommerce/api-utils/encodeOpaqueId.js";
+import decodeOpaqueIdForNamespace from "@reactioncommerce/api-utils/decodeOpaqueIdForNamespace.js";
+import importAsString from "@reactioncommerce/api-utils/importAsString.js";
+import Factory from "/tests/util/factory.js";
+import TestApp from "/tests/util/TestApp.js";
+
+const createTaxRateMutation = importAsString("./createTaxRateMutation.graphql");
+const updateTaxRateMutation = importAsString("./updateTaxRateMutation.graphql");
+const deleteTaxRateMutation = importAsString("./deleteTaxRateMutation.graphql");
+
+jest.setTimeout(300000);
+
+let createTaxRate;
+let mockAdminAccount;
+let deleteTaxRate;
+let shopId;
+let shopOpaqueId;
+let testApp;
+let taxRateOpaqueId;
+let updateTaxRate;
+
+beforeAll(async () => {
+  testApp = new TestApp();
+  await testApp.start();
+  shopId = await testApp.insertPrimaryShop();
+  createTaxRate = testApp.mutate(createTaxRateMutation);
+  updateTaxRate = testApp.mutate(updateTaxRateMutation);
+  deleteTaxRate = testApp.mutate(deleteTaxRateMutation);
+
+  mockAdminAccount = Factory.Account.makeOne({
+    _id: "mockAdminAccount",
+    roles: {
+      [shopId]: ["admin", "shopManagerGroupPermission", "someOtherPermission", "customerGroupPermission"]
+    },
+    shopId
+  });
+  await testApp.createUserAndAccount(mockAdminAccount);
+
+  shopOpaqueId = encodeOpaqueId("reaction/shop", shopId);
+});
+
+afterAll(async () => {
+  await testApp.collections.TaxRates.deleteMany({});
+  await testApp.collections.Shops.deleteMany({});
+  await testApp.stop();
+});
+
+test("user can add a tax rate", async () => {
+  await testApp.setLoggedInUser(mockAdminAccount);
+
+  const taxRateInput = {
+    shopId: shopOpaqueId,
+    region: "CA",
+    rate: 0.10,
+    country: "USA",
+    postal: "90066",
+    taxCode: "CODE"
+  };
+
+  let result;
+  try {
+    result = await createTaxRate(taxRateInput);
+  } catch (error) {
+    expect(error).toBeUndefined();
+    return;
+  }
+
+  const { _id: createdTaxRateOpaqueId, ...createdTaxRate } = result.createTaxRate.taxRate;
+
+  // Save this for the next tests for updating and deleting;
+  taxRateOpaqueId = createdTaxRateOpaqueId;
+
+  // Validate the response
+  // _id is omitted since the ID is tested for proper opaque ID conversion in the DB test below.
+  const expectedTaxRateResponse = {
+    shop: {
+      _id: shopOpaqueId
+    },
+    region: "CA",
+    rate: 0.10,
+    country: "USA",
+    postal: "90066",
+    taxCode: "CODE"
+  };
+
+  expect(createdTaxRate).toEqual(expectedTaxRateResponse);
+
+  // Check the database for the new TaxRate document
```

Comment:
is this double-check really necessary? the GraphQL operation returns the created/updated document, seems redundant to query the db, unless I am missing something?

Reply:
The GQL mutation/query returns a resolved document, that may not be what's actually in the database. For example, the field `shop` is available in queries, but only `shopId` is stored on the document in the DB. A resolver may add additional fields or formatting to a field.

The query to the database is to ensure the document was stored how it's expected to be before any GQL resolvers manipulate it.

Reply:
make sense